File: LeetcodeProblems/blind-75/matrix/set-matrix-zeroes/solution.py
# Set Matrix Zeroes - LeetCode #73

from typing import List

class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        if not matrix or not matrix[0]:
            return
        
        m = len(matrix)
        n = len(matrix[0])
        first_row_zero = False
        first_col_zero = False
        
        # Check first row
        for j in range(n):
            if matrix[0][j] == 0:
                first_row_zero = True
                break
        
        # Check first col
        for i in range(m):
            if matrix[i][0] == 0:
                first_col_zero = True
                break
        
        # Mark
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
        
        # Set zeros
        for i in range(1, m):
            for j in range(1, n):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0
        
        # Set first row
        if first_row_zero:
            for j in range(n):
                matrix[0][j] = 0
        
        # Set first col
        if first_col_zero:
            for i in range(m):
                matrix[i][0] = 0

File: LeetcodeProblems/blind-75/matrix/set-matrix-zeroes/solution.js
// Set Matrix Zeroes - LeetCode #73

/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var setZeroes = function(matrix) {
    if (matrix.length === 0) return;
    
    const m = matrix.length;
    const n = matrix[0].length;
    let firstRowZero = false;
    let firstColZero = false;
    
    // Check first row
    for (let j = 0; j < n; j++) {
        if (matrix[0][j] === 0) firstRowZero = true;
    }
    
    // Check first col
    for (let i = 0; i < m; i++) {
        if (matrix[i][0] === 0) firstColZero = true;
    }
    
    // Mark
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            if (matrix[i][j] === 0) {
                matrix[i][0] = 0;
                matrix[0][j] = 0;
            }
        }
    }
    
    // Set zeros
    for (let i = 1; i < m; i++) {
        for (let j = 1; j < n; j++) {
            if (matrix[i][0] === 0 || matrix[0][j] === 0) {
                matrix[i][j] = 0;
            }
        }
    }
    
    // Set first row
    if (firstRowZero) {
        for (let j = 0; j < n; j++) {
            matrix[0][j] = 0;
        }
    }
    
    // Set first col
    if (firstColZero) {
        for (let i = 0; i < m; i++) {
            matrix[i][0] = 0;
        }
    }
};
